src/backend/src/yaksuri_progress.c

```diff
@@ -0,0 +1,532 @@
+
+    assert(need_host_buf || need_device_buf);
+
+    *subop = NULL;
+
+    /* figure out if we actually have enough buffer space */
+    if (need_device_buf) {
+        uintptr_t d_nelems;
+        if (yaksuri_global.cuda.device.slab_head_offset == 0 &&
+            yaksuri_global.cuda.device.slab_tail_offset == 0) {
+            d_nelems = TMPBUF_SLAB_SIZE / elem->pup.type->size;
+            device_tmpbuf_offset = yaksuri_global.cuda.device.slab_tail_offset;
+        } else if (yaksuri_global.cuda.device.slab_tail_offset >
+                   yaksuri_global.cuda.device.slab_head_offset) {
+            uintptr_t count =
+                (TMPBUF_SLAB_SIZE -
+                 yaksuri_global.cuda.device.slab_tail_offset) / elem->pup.type->size;
+            if (count) {
+                d_nelems = count;
+                device_tmpbuf_offset = yaksuri_global.cuda.device.slab_tail_offset;
+            } else {
+                d_nelems = yaksuri_global.cuda.device.slab_head_offset / elem->pup.type->size;
+                device_tmpbuf_offset = 0;
+            }
+        } else {
+            d_nelems =
+                (yaksuri_global.cuda.device.slab_head_offset -
+                 yaksuri_global.cuda.device.slab_tail_offset) / elem->pup.type->size;
+            device_tmpbuf_offset = yaksuri_global.cuda.device.slab_tail_offset;
+        }
+
+        if (nelems > d_nelems)
+            nelems = d_nelems;
+    }
+
+    if (need_host_buf) {
+        uintptr_t h_nelems;
+        if (yaksuri_global.cuda.host.slab_head_offset == 0 &&
+            yaksuri_global.cuda.host.slab_tail_offset == 0) {
+            h_nelems = TMPBUF_SLAB_SIZE / elem->pup.type->size;
+            host_tmpbuf_offset = yaksuri_global.cuda.host.slab_tail_offset;
+        } else if (yaksuri_global.cuda.host.slab_tail_offset >
+                   yaksuri_global.cuda.host.slab_head_offset) {
+            uintptr_t count =
+                (TMPBUF_SLAB_SIZE -
+                 yaksuri_global.cuda.host.slab_tail_offset) / elem->pup.type->size;
+            if (count) {
+                h_nelems = count;
+                host_tmpbuf_offset = yaksuri_global.cuda.host.slab_tail_offset;
+            } else {
+                h_nelems = yaksuri_global.cuda.host.slab_head_offset / elem->pup.type->size;
+                host_tmpbuf_offset = 0;
+            }
+        } else {
+            h_nelems =
+                (yaksuri_global.cuda.host.slab_head_offset -
+                 yaksuri_global.cuda.host.slab_tail_offset) / elem->pup.type->size;
+            host_tmpbuf_offset = yaksuri_global.cuda.host.slab_tail_offset;
+        }
+
+        if (nelems > h_nelems)
+            nelems = h_nelems;
+    }
+
+    if (nelems > elem->pup.count - elem->pup.completed_count - elem->pup.issued_count)
+        nelems = elem->pup.count - elem->pup.completed_count - elem->pup.issued_count;
+
+    /* if we don't have enough space, return */
+    if (nelems == 0) {
+        goto fn_exit;
+    }
+
```

Comment:
Please remove double empty lines.

Reply:
The double empty lines are intentional to separate logically different parts of the code.